modules/javafx.graphics/src/main/java/com/sun/glass/ui/mac/MacApplication.java

```diff
@@ -63,6 +63,28 @@ final class MacApplication extends Application implements InvokeLaterDispatcher.
     private boolean isTaskbarApplication = false;
     private final InvokeLaterDispatcher invokeLaterDispatcher;
 
+    private static final CountDownLatch keepAliveLatch = new CountDownLatch(1);
+
+    private static void startKeepAliveThread() {
+        // Start a non-daemon thread to keep the JavaFX toolkit alive
+        // This is necessary on macOS because the JavaFX Application thread
+        // is attached to the macOS Appkit Thread as a daemon thread
+        Thread thr = new Thread(() -> {
+            try {
+                keepAliveLatch.await();
+            } catch (InterruptedException ex) {
+                throw new RuntimeException("Unexpected exception: ", ex);
+            }
+        });
+        thr.setName("JavaFX-KeepAlive");
+        thr.setDaemon(false);
+        thr.start();
```

Comment:
is it possible to reuse some other thread for this purpose?
this may not be an issue when virtual threads take hold, but right now threads are precious native resource.

Reply:
No. The whole reason we are doing this is that there are no other non-daemon threads. I updated the docs to make this more clear. Anyway, threads aren't _that_ scarce as long as you don't create hundreds of them.

Loosely related: This reminds me that `javafx.application.Application.launch` does create an unnecessary thread (on all platforms). There is a long-standing issue to get rid of it:

[JDK-8090323](https://bugs.openjdk.org/browse/JDK-8090323): Consider eliminating the JavaFX-Launcher Thread

It isn't high priority, but would be good cleanup.

Reply:
thank you for clarification!
 
never underestimate the happiness with which the app and library developers create new threads...